[WIP] The Patch PR

Cheatoid opened the pull request:
I touched several files around here.
Note, while I have made some fixes initially, but I am still not done with this PR entirely. This is WIP.
New commits I make from now on will be smaller in size, to help with rebasing/merging if you need to.

Where to begin? I tried to comment on most of the stuff that needed comment.
If you go over the code you can see the `MYTODO` and some `TODO` marks too. I will try to address the `MYTODO` within this PR.
So, that being said, don't merge this in, yet. Or actually, I don't know, if you treat your `master` as a stable or development branch...

I just want you to give it a quick look at the changes so you can adapt your code sooner.
You can place comments on any line in the diff and I can explain what am I doing, if you want me to.
For now, I guess I can explain in comments about what is going on, and what exactly am I doing. (Comments incoming shortly.)

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_coroutines.lua:
```diff
@@ -24,53 +24,56 @@ vex.registerExtension("coroutines", false, "Allows E2s to use coroutines.")
 -- Coroutine Object handling
 
 registerType("coroutine", "xco", nil,
-	nil,
```
No code is changed.
I think we should be using spaces for indentation.
These are tabs, so I've reformatted all the touched files to use spaces.
Be sure to configure your editor properly, so this mix doesn't happen again. (I use Visual Studio Code and Sublime Text 4.)

Vurv78 replied:
I was actually using spaces before seeing this, do you mean using tabs instead? I had no idea vscode auto-converted tabs to spaces, shows how much of a noob I am. Switched to tabs for now

Cheatoid replied:
No. Use spaces.

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_coroutines.lua:
```diff
@@ -24,53 +24,56 @@ vex.registerExtension("coroutines", false, "Allows E2s to use coroutines.")
 
 registerCallback("destruct",function(self) -- On e2 deleted, deletes all coroutines. gives compiler to function
-    self.coroutines = {}
+    self.coroutines = {} -- TODO: Why not set this to nil?
```
Should this be set to `nil` instead?

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_coroutines.lua:
```diff
@@ -24,53 +24,56 @@ vex.registerExtension("coroutines", false, "Allows E2s to use coroutines.")
     return lhs == rhs
 end
 
+e2function number operator!=(coroutine lhs, coroutine rhs) -- if(coroutineRunning()!=Co)
```
- [x] Added `!=` operator for `coroutine`.

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_coroutines.lua:
```diff
@@ -80,37 +83,24 @@ local function loadPrfData(instance,data)
     if compiler.coroutines[thread] then return thread end
 end
 
-local function getE2FuncFromStr(compiler,funcname)
```
- [x] Refactored this function and moved it into a single place, [click here to see](https://github.com/Vurv78/VExtensions/pull/15/files#diff-13513553927461b66e8fd31dbdf4979c1444325be4f59759e4217ca11732bd96R153).

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_coroutines.lua:
```diff
@@ -149,41 +141,41 @@ e2function void coroutineWait(n)
-    return thread
+-- Returns the actual thread being run. No need to return a number before since we added the if operator on coroutines.
+e2function coroutine coroutineRunning()
+    return runningCo(self)
```
This code works exactly like before, just simplified it.

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_rune2.lua:
```diff
@@ -1,5 +1,6 @@
 local table_remove = table.remove
 local table_insert = table.insert
+local luaTableToE2, getE2UDF, getE2Func = vex.luaTableToE2, vex.getE2UDF, vex.getE2Func
```
Import the (refactored) helper functions which we will use in this file.

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_rune2.lua:
```diff
@@ -20,69 +21,54 @@ local function buildBody(args) -- Why does the wireteam do this
-local function runE2InstanceSafe(compiler,func,body) -- Varargs to pass to the e2 function
-    -- Will always return pcallerror,errstr first even if it didn't error.
-    local args = {pcall(func,compiler,body)}
+local function runE2InstanceSafe(compiler,func,body,...)
```
I will try to add vararg support for the `try`, so this is just the initial part of it I guess.

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_rune2.lua:
```diff
@@ -20,69 +21,54 @@ local function buildBody(args) -- Why does the wireteam do this
     end
+    -- If unsuccessful, don't return args, that would be a waste
+    local errmsg = table_remove(args,1)
+    if errmsg == "exit" then return true,nil,{} end -- nice exit()
```
I have changed the string `"none"` to `nil` since it is not really used.

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_rune2.lua:
```diff
@@ -20,69 +21,54 @@ local function buildBody(args) -- Why does the wireteam do this
+    return false,errmsg
 end
 
-local function getE2FuncFromStr(compiler,funcname)
```
Same as earlier, this function has been refactored and moved into the single place, into helpers file.
Also, I forgot to say, I have also fixed that pattern as well.
It was failing to find E2 UDF if the name contained underscore(s), for example `h_3llo_` is a valid UDF name.

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_rune2.lua:
```diff
@@ -20,69 +21,54 @@ local function buildBody(args) -- Why does the wireteam do this
-        return e2func
-    end
-end
+-- TODO: Set E2 costs ( __e2setcost(N) ).
```
You forgot set E2 cost for these functions. We can determine the costs and fix this later. This is just a reminder.

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_rune2.lua:
```diff
@@ -20,69 +21,54 @@ local function buildBody(args) -- Why does the wireteam do this
-    local success,errstr,args = runE2InstanceSafe(self,tryfnc)
+-- Returns table, first argument is a number stating whether the function executed successfully, rest are varargs.
+-- TODO: variadic/varargs support [see array(...) / table(...) / select(N...) E2 functions for reference]
+e2function table try(string try)
```
I have changed the return type from `array` to `table`, because using an `array` was a really bad design decision. Here's why:
[E2 array **can not** hold another array/table/gtable]() and that's the problem, because this means if you were to call your UDF which returns a `table` value, then you won't be able to access that table.
- [x] Fix the `try` function. Change its return type from `array` to `table` to allow arrays/tables to be returned from UDF.

Cheatoid replied:
Also, such a function should not be throwing an error by design!
- [x] Made `try` not throw an error if the UDF is not defined.

Vurv78 replied:
Originally I made this function in mind of the defined() func, since you could just check for yourself. Ig this is fine

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_rune2.lua:
```diff
@@ -20,69 +21,54 @@ local function buildBody(args) -- Why does the wireteam do this
 
 
--- Literally like xpcall()
+-- Literally like xpcall() ... Actually, not really, no it is not; If you get/make an error in your E2 catch callback, it will stop executing...
```
I guess I already commented on why this should be removed.

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_selfaware2.lua:
```diff
@@ -1,54 +1,125 @@
+  / ___/ ___   / // __/  /   | _      __ ____ _ _____ ___     |__ \
+  \__ \ / _ \ / // /_   / /| || | /| / // __ `// ___// _ \    __/ /
+ ___/ //  __// // __/  / ___ || |/ |/ // /_/ // /   /  __/   / __/
+/____/ \___//_//_/    /_/  |_||__/|__/ \__,_//_/    \___/   /____/
```
There were two `r` in there.

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_selfaware2.lua:
```diff
@@ -1,54 +1,125 @@
 ]]
 
--- Reused from coroutine-core
-local function getE2UDF(compiler,funcname)
```
Same as earlier, this has been refactored and moved into the helpers.
(Same for the `getE2Func` below.)
We import these functions here so we can use them (line 15).

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_selfaware2.lua:
```diff
@@ -1,54 +1,125 @@
-        return e2func[3]
-    end
-end
+-- TODO: Set E2 costs ( __e2setcost(N) ).
```
Another file, another reminder about determining and setting E2 costs, later...

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_selfaware2.lua:
```diff
@@ -1,54 +1,125 @@
-end
+-- TODO: Set E2 costs ( __e2setcost(N) ).
 
--- Ex: print(defined("print(...)")) or print(defined("e:health()"))
```
Just a note, you don't specify `e:` like that, it has to be after the `(` parenthesis.

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_vex_main.lua:
```diff
@@ -7,53 +7,32 @@
 
-local E2Table = function() return {n={},ntypes={},s={},stypes={},size=0} end
-
-local function luaTablToE2(T)
```
Refactored, moved into helpers, [click here to see](https://github.com/Vurv78/VExtensions/pull/15/files#diff-13513553927461b66e8fd31dbdf4979c1444325be4f59759e4217ca11732bd96R41).

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_selfaware2.lua:
```diff
@@ -1,54 +1,125 @@
+end
+
+-- Returns an array containing only names of all User-Defined Functions.
+e2function array udfNames()
```
New E2 functions down below, `udfNames` seems complete and `udfAll` is being worked on...

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_selfaware2.lua:
```diff
@@ -1,54 +1,125 @@
+    -- Populate keys on the table (to avoid duplicate entries) and then get the keys on return.
+    local ret = {}
+    for name in pairs(self.funcs_ret) do
+        local idx = string_find(name, "(", 2, true) -- This should never return a nil.
```
Not using patterns here because I think this will be faster; the 4th argument `true` disables patterns for performance.

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_vex_main.lua:
```diff
@@ -7,53 +7,32 @@
 __e2setcost(10)
 e2function table rangerOffsetManual(vector pos,vector endpos, array filt)
-    local Start = Vector(pos[1],pos[2],pos[3])
-    local End = Vector(endpos[1],endpos[2],endpos[3])
```
Simplified this piece of code.

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_vex_main.lua:
```diff
@@ -7,53 +7,32 @@
     if #filter == 0 then self.data.rangerfilter = {} return 1 end
     if #filter > 3000 then return 0 end
     self.prf = self.prf + #filter*1.5
-	local fixed = {}
```
Tabs go brrrrrrrrrrrrrrrrr.

Cheatoid commented on lua/vex_library/modules/server/e2helperfuncs.lua:
```diff
@@ -16,19 +16,89 @@ vex.registerExtension = function(name,enabled,helpstr,...)
     E2Lib.RegisterExtension(name,enabled,helpstr,...)
 end
 
+vex.registerConstant = function(name, value, ...)
```
- [x] Added the `vex.registerConstant`. A small utility to help with syntax while registering E2 constant at the same time.

Cheatoid commented on lua/vex_library/modules/server/e2helperfuncs.lua:
```diff
@@ -16,19 +16,89 @@ vex.registerExtension = function(name,enabled,helpstr,...)
+
+local EMPTY_TABLE = vex.newE2Table()
+local istable, type = istable, type
+vex.isE2Table = function(tbl)
```
- [x] Added `vex.isE2Table`, another utility function.
It is used in the new `vex.luaTableToE2` and existing `vex.getE2Type` functions to determine whether the given `tbl` is an E2 table.

Cheatoid commented on lua/vex_library/modules/server/e2helperfuncs.lua:
```diff
@@ -16,19 +16,89 @@ vex.registerExtension = function(name,enabled,helpstr,...)
+end
+
+local isnumber, isstring, string_lower, table_IsSequential = isnumber, isstring, string.lower, table.IsSequential
+vex.luaTableToE2 = function(tbl,arrayOptimization) -- TODO: Fix cyclic/infinite Lo0oP (see GMod's PrintTable code for reference).
```
It is possible for this function to enter an infinite cyclic loop, as the line comment suggests.
[Here's the source code of the `PrintTable`](https://github.com/Facepunch/garrysmod/blob/master/garrysmod/lua/includes/util.lua#L58-L92), see how it keeps the track of *already done tables*...

Cheatoid commented on lua/vex_library/modules/server/e2helperfuncs.lua:
```diff
@@ -48,7 +118,7 @@ end
 vex.callE2Hook = function(id,callback,...)
     -- Executes all of the chips in the e2 hook.
     local runs = runs_on[id]
-    for chip,_ in pairs(runs) do
```
Removed that unused variable.

Cheatoid commented on lua/vex_library/modules/server/e2helperfuncs.lua:
```diff
@@ -75,4 +145,39 @@ vex.createE2Hook = function(hookname,id,callback)
+end
+
+
+local E2FuncNamePattern = "^([a-z][a-zA-Z0-9_]*)%("
```
And here's the fixed pattern along the refactored functions that are now in one file, one place, much easier to maintain 😄

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_selfaware2.lua:
```diff
@@ -1,54 +1,125 @@
+    if funcDirect then return 1 end -- Builtin perfect match.
+    -- Name only match after this point :(
+    -- Which one to prefer if they are both defined with identical name? It makes sense to prefer UDF?    *smh*
+    -- Should this return a negative number to indicate uncertainity?
```
No no...
I think you misunderstand/misread. By the way, you **can** redefine existing E2 functions (only applies to variadic/vararg ones).
Here is a code example to help understand what exactly I meant:
```golo
function void print(Text:string) # Didn't know this? Now you do ;)
{
    print("[info]", Text) # Yes this will call into the builtin print because of the signature/overloading.
}

print(defined("print")) # What do you expect to get from this?
# Since there are 2 of them now, builtin and UDF, both called with the same name.

# More examples:
function applyForce(Ar:array, V:vector)
{
    # Omitted for clarity.
}

print(defined("applyForce"))  # Now again, what do you expect this to say?
```

Cheatoid replied:
And also of course, you can define your own functions such as `applyForce` (non-variadic) too, as long as the entire UDF signature is *not* already defined.

Vurv78 replied:
Splitting them doesn't sound bad, just keep the name of 'defined' as it is, it will handle e2functions and 'definedUDF' will be for udfs

Cheatoid replied:
Could make it overloaded, just like I do for the `udfAll`.
By introducing a number argument (that would be either 0 or 1) to indicate which one to prefer?
So the `defined`'s signature would changed to:
`e2function number defined(string funcname, number preferBuiltin)`
And then a number argument would be checked for (only when it is not certain), do you want me to do this?
Note, changing argument's signature would be a breaking change. (So, how many users are out there using `defined`?)

Vurv78 replied:
I feel that going with defined vs definedUDF would be a bit more intuitive than making more of these multi-selectional functions, but if you really wanted to do the preferBuiltin route, please keep the number arg optional so it isn't a breaking change like you said (Well it'd still be different in that it returns only e2functions and not udfs)

Cheatoid replied:
But, hah, rethinking this. You could also use the new `udfNames` / `udfAll` to specifically check for UDF.
So, I believe that the order should be swapped now, it should prefer builtin instead of UDF.

Cheatoid commented:
<details>
  <summary>Just a quick test to confirm the previous commit (expand me to see code)</summary>
  
```golo
function void print(Text:string) # Didn't know this? Now you do ;)
{
  # Yes this will call into the builtin print because of the signature/overloading.
  print("[info]", Text)
}

print("defined(\"print(...)\"): " + defined("print(...)")) # This will obviously always give 1.
print("defined(\"print\"): " + defined("print")) # Now prefers builtin, so it should give 1.
# Since there are 2 of them now, builtin and UDF, both called with the identical name.

# Another example:
function applyForce(V:vector, Ar:array) { }

print("defined(\"applyForce(v)\"): " + defined("applyForce(v)")) # Obviously, should always give 1.
print("defined(\"applyForce\"): " + defined("applyForce")) # And this again, should give 1.

function t_3s_t() { }
print("defined(\"t_3s_t\"): " + defined("t_3s_t")) # Obviously, should always give 2.
```
  > ![image](https://user-images.githubusercontent.com/13347909/100835578-2c58dc00-346e-11eb-950f-a6e29ff8f161.png)
</details>

Cheatoid commented:
<details>
  <summary>A quick test for the newly introduced functions `udfNames` and `udfAll` (click me)</summary>

```golo
function void foo(Num, Text:string){}
function string entity:myFunc(Col:vector4, Ar:array){return""}
function table entity:myFunc(Rot:angle, Pos:vector){return table()}

print("[ udfNames() ]")
printTable(udfNames())
print("[ udfAll(_UDF_ALL_FLAT) ]")
printTable(udfAll(_UDF_ALL_FLAT))
print("[ udfAll(_UDF_ALL_DNC) ]")
printTable(udfAll(_UDF_ALL_DNC))
```

> ![image](https://user-images.githubusercontent.com/13347909/100838862-84461180-3473-11eb-8730-532d26a7f4a1.png)
</details>

Almost there...

Cheatoid commented:
@Vurv78 Should I remove the `udfNames` function?
Since `udfAll` can be used (in *D&C* mode), followed by calling `:keys()` to obtain all the UDF names:
```golo
function func() { }

printTable( udfNames() )
# udfNames is kind of shorthand for:
printTable( udfAll( _UDF_ALL_DNC ):keys() )
```
Both would give the same table result, except the `:keys()` call does add some dynamic cost tho.

So, remove or leave it as is?

Vurv78 commented:
I'd say remove it, there's pretty much no instance where you're going to use this enough times for the difference in cpu time to matter. Also, I feel that udfAll might be a little weird as to what it gives you, maybe it should be called getUDFs? Or getUDFDatas/Infos? I like the style of get/set*

Cheatoid commented:
Uhhhhh, breaking change ~.~

Vurv78 commented:
Is that everything to do ?

Cheatoid commented:
No?

Vurv78 commented:
I pressed a button 👍

Cheatoid commented:
The functions relied on the array optimization as an opt-in, why took it away?
And I have more commits to push but i cant now..

Vurv78 commented:
Adding arrayOpt back I forgot we're using IsSequential here, which would be terrible for your getbuiltins function

Cheatoid commented:
btw `vex.luaTableToE2` is still vulnerable to infinite/cyclic loop.

Vurv78 commented:
How? Also I just realised I messed up a ton...

Cheatoid commented:
> How?

Oh actually, only if you were to do it from Lua, an E2 table should just be passed over as is.

Cheatoid commented:
E2 type ID is either 1 char, or 3 chars long when the first char is `x`.

By the way, there is an invert table which maps typeids to full typenames (uppercased): `wire_expression_types2` (notice `2` at the end).
idk what are you doing, but you could use that to avoid having to change printglobal code I suppose.

Cheatoid commented:
I am just about to add in one more function in the selfaware2, and then it would have been ready for review, if just it weren't broken.. (¬_¬)

Cheatoid commented on lua/vex_library/modules/server/e2helperfuncs.lua:
```diff
@@ -84,8 +85,15 @@ vex.luaTableToE2 = function(tbl)
+            elseif arrayOptimization and table.IsSequential(V) then
+                t[K],t_types[K] = V,"r"
+            else
+                t[K] = vex.luaTableToE2(V)
```
Not passing it into recursion?

Vurv78 replied:
wdym

Cheatoid replied:
The flag is lost after the first depth. For example if you have a table and another table inside that is supposed to have arrays within it, then it would treat it as a table and not an array. I will fix this.

Vurv78 replied:
Agh I'm blind, thanks

Cheatoid commented:
We could ignore the missing parts for now, these can be implemented with another PR (missing entries for the `getUserFunctionInfo`).
Vararg support for `try` and coroutine, also within another PR.

----

- [ ] Perhaps "merge" the `getFunctionPath` into the `getBuiltinFuncInfo` as a new entry in the table? Since it is more verbose version, a function path could easily be added as a new entry in the array.
- [ ] Write E2 docs (for new functions).

Cheatoid commented:
I just realized how you had messed up helper functions...  :/
> We don't need to sanitize anything else like vectors or angles.

So let me explain why `Angle` and `Vector` have to be sanitized, even Wire has this vulnerability (in several places, but fu** them, i ain't fixing their code), for example:
```lua
local DummyAng = Angle(1, 2, 3)
e2function angle dummyAng()
    return DummyAng -- This code is vulnerable to reference modification!
end
```
What could go wrong, you would think it would return a copy of the angle at the time of return, well it doesn't. Because `Angle `and `Vector` are bound to GLua as "classes", this means they are like Lua table, which is pass-by-reference rather than pass-by-value.
Which practically means:
```golo
# my, malicious E2 code:
Ang = dummyAng()
print( "Initial: " + Ang:pitch() )
Ang[1] = 1337
print( "First time: " + Ang:pitch() )

# someone else's, another E2 code:
Ang = dummyAng() # This will fetch the modified reference!
print( "Second time: " + Ang:pitch() )
```
![image](https://user-images.githubusercontent.com/13347909/101088017-5962d700-35b3-11eb-8085-396463ff0746.png)

Cheatoid commented:
Seems like array optimization is not working as expected.

Cheatoid commented:
```golo
function string entity:myFunc(Col:vector4, Ar:array){return""}
function table entity:myFunc(Rot:angle, Pos:vector){return table()}
printTable(getUserFunctionInfo(_UDF_DNC)["myFunc", table]:typeids())
```
> 1    =    t
> 2    =    t

This should be printing out `r` not `t`

Vurv78 commented:
I don't think we should be hardcoding this much just to get e2 types.
```
    if isnumber(v) then return "n" end
    if isbool(v) then return "n", true end -- Returning an additional true, to indicate it must be sanitized (using `vex.sanitizeLuaVar`).
    if isstring(v) then return "s" end
    if isentity(v) then return "e" end
```
This is why we loop through [6] in wire_expression2_types, because all of these types already have a predefined function that tells you what they are. The only exception should be for things like lua angles, since once again with e2's lame compiler and no 100% sure way to retrieve the type, a vector should always be returned since it has the same criteria as an angle in e2.

See: https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/vector.lua#L30
https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/angle.lua#L10
https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/bone.lua#L74

Also coincidentally in vector.lua we can see that e2 sometimes passes vectors into the code which I see no problem with

In response to wrapping angles and vectors... why would we need to do this? The developer should know not to just pass table-like variables used in lua to e2 that can be set... The same example you gave with the angle works if you pass a table.

Thinking on it more, arrayOptimization should be set by default, it shouldn't just be an option. I don't think there's any case where you'd *want* to pass every lua table as an e2 table in there. However, if we do make it by default, we should check if there's any tables in the table and if so avoid that
Could probably make a vex.isE2Array function for that that just once again checks if it is sequential but also looks for any tables in it.

vex.getE2Type should return the typeid rather than the actual type name, since that's what e2 uses to define types and look for them in tables. I'm not sure why you reverted it to use the name again

TypeName = wire_expression_types2[TypeData[1]][1] Just returns TypeData[1] btw
And you're returning conflicting things, return TypeName == "NORMAL" and "NUMBER" or TypeName

Which would either return NUMBER if it is a normal or "r","a" etc

Vurv78 commented:
https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/matrix.lua#L381

Agh this kinda stuff makes me wanna work on starfall, hopefully this will be the only time we have to deal with this type cancer

Cheatoid commented:
> This is why we loop through [6] in wire_expression2_types, because all of these types already have a predefined function that tells you what they are

Except there are 2 problems  with that, it might return a wrong type id, such as `xgt` which is the global/shared table type...
Therefore we cant rely on the [6] thing, it is made to be called the last in that chain. Another problem was... dam i forgot it.
But yeah, can't rely on this function. Also it is slower than just checking it ourselves.

Vurv78 commented:
Also I don't think we should guess for anything that doesn't have an explicit lua type or anything, cause otherwise that's *actually* guessing and despite the name of the function (I'm stupid) I want it to be 100% sure, so In this case everything would be tables or arrays except for things like angles and lua matrices which would be a matrix4 (I think)

Cheatoid commented:
GLua VMatrix has nothing to do with E2 matrix*.

Vurv78 commented:
Ah well, it's not like anyone uses e2 matrices anyway, pretty much useless

Cheatoid commented:
Yeah right, so I dont understand why did you make them as globals?
The only place where this is actually used is selfaware2, and it only requires number/string along with array support, nothing less nothing more.
I am already writing a more localized version of this function, it will be as fast as possible because I prefer speed in there.
So, do whatever you want with helper stuff, get rid of it, fix it, do whatever you want. I am not touching the helper file any more.
Let's just get this back on track instead of having to mess with impossible type-system shi*

Vurv78 commented:
Yeah, I'll get rid of them being global, idr why I made them global when they are literally only used by getE2Type, but here we are.. and yeah, just want to make sure that this is basically finalized so we don't have to touch it again, your input has been super useful, so thanks dude. Also, looking forward to if you ever add any mini-extensions to here

Cheatoid commented:
> TypeName = wire_expression_types2[TypeData[1]][1] Just returns TypeData[1] btw
> And you're returning conflicting things, return TypeName == "NORMAL" and "NUMBER" or TypeName

> Which would either return NUMBER if it is a normal or "r","a" etc

I had reverted that back, so it is "STRING", "ARRAY".. etc instead, I reverted the breaking change because that would also broke other addons outside. So why make breaking change if there's no need for it.
The second table is the inverted one, so it should return full type name (uppercased). Not the type id.
Are you looking at the most recent commit?

Vurv78 commented:
TypeName would be the full type name in that case if you wanted to go about it, what I was trying to say was that I didn't understand why you were returning the uppercase type names rather than the type ids for the table, because this would be used directly by luaTableToE2 so it gets types

And yeah idk why 'normal' still exists... I don't know why the wireteam wants to pull a Windows :tm: and keep features nearly a decade old

Cheatoid commented:
> TypeName would be the full type name in that case if you wanted to go about it, what I was trying to say was that I didn't understand why you were returning the uppercase type names rather than the type ids for the table

Well I just want to avoid breaking changes. This is how you had it as, and that's what I reverted back:
<https://github.com/Vurv78/VExtensions/pull/15/commits/7326d5533e52b3d65c2daede9509b965cd86577a>

See your current master:
<https://github.com/Vurv78/VExtensions/blob/master/lua/vex_library/modules/server/e2helperfuncs.lua#L30>
<https://github.com/Vurv78/VExtensions/blob/master/lua/entities/gmod_wire_expression2/core/custom/sv_printglobal.lua#L47>

I reverted your changes you made in this PR with this commit:
<https://github.com/Vurv78/VExtensions/pull/15/commits/887a7668b5284ffd0f6896f9290bc6591990776e#diff-13513553927461b66e8fd31dbdf4979c1444325be4f59759e4217ca11732bd96L100>

> because this would be used directly by luaTableToE2 so it gets types

Again, you can't use the [6] reliably at all... it **will** give you the wrong shit.
It is **undefined behavior** if you call the [6] *before* doing manual checks in that chain.
Calling [6] may return `t`, but is NOT guaranteed to do so, because it can also return the `xgt` (or any other), for example:
`lua_run print( vex.getE2Type( {foo="bar"}, true ) )`
![image](https://user-images.githubusercontent.com/13347909/101116994-6816b180-35e6-11eb-9d78-0588be347c31.png)
> `GTABLE`

See... That's a completely wrong E2 type.
<https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/globalvars.lua#L23>

Cheatoid commented:
> I don't think we should be hardcoding this much just to get e2 types.

That's exactly what we should be doing tho. Because builtin E2 type-system isn't going to change lol. Remember `normal` ? xD

Vurv78 commented:
Well yeah, that's why we have the guess function in the first place, I was just talking about the commented out matrix stuff and the basic types like numbers and strings, I suppose it'd be faster to keep those though instead of looping through ones that might not be true until it lands on string/number.

Cheatoid commented:
Just ignore that commented out stuff, nothing requires that, that's why I commented it out. You could do that, but it would be a waste on the CPU time... Don't add support for values you're not even going to be using in VExtensions.

Cheatoid commented:
> I'll push something in a bit

Do whatever you want in helpers, I am making a localized and specialized version just for use in selfaware2, will be done in a moment.
After that, I'd like to go for a final review and to merge this, as this page is getting big and slow to load.
Everything else can be then continued upon in the second PR..

Cheatoid commented:
By the way, about the array optimization flag, if you change that to be *always-enabled*, that will make performance worse, because then it will always end up calling `table.IsSequential` even if you know there won't be any arrays.
For example if you just give it a table with string keys. That would make time complexity much worse if it is always enabled.
That's why I made it as an opt-in. Let the developer decide whether it should attempt to optimize into an array, and then you only end up calling `table.IsSequential` only if the developer has opted-into it for their particular table.
Forcing such thing is a bad thing.
But, do whatever you want, as long as you just pass E2 tables as they are, so you don't break the `try` E2 function..
I am fine with it. I don't care about helpers anymore.

Cheatoid commented:
The specialized `luaTableToE2` for SelfAware2 has also addressed the bug with array optimization flag not working as expected previously (I couldn't figure out why it didn't work and I stopped caring about the `vex.luaTableToE2` at that point).
```golo
function string entity:myFunc(){return""}
function string entity:myFunc(A){return""}
FN = getUserFunctionInfo(_UDF_DNC)[ "myFunc", table ]
print( "[FuncInfos]" ) printTable( FN )
print( "[TypeIDs]" ) printTable( FN:typeids() )
```
The above now correctly outputs the desired table structure, where function info is stored inside an array -- not a table.
```
[FuncInfos]
1:
		2	=	s
		1	=	e:n
2:
		2	=	s
		1	=	e:
[TypeIDs]
1	=	r
2	=	r
```

Notice the `r` instead of `t` in the `[TypeIDs]`. Better performance + bug squash at the same time.
Perfection. 😄

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_selfaware2.lua:
```diff
@@ -131,17 +174,19 @@ local GET_UDF_MODE = {
+                -- TODO (Patch #2): entries [3] and [4] in the table (not sure yet how to obtain these)
                 collection[#collection + 1] = { [1] = sig, [2] = returnType or "" }
             end
-            return luaTableToE2(res, false) -- Convert to E2 table *without* array optimization.
```
Oh riight, it was never enabled 😆
But even if you enabled it, that would have produced an incorrect table layout (it would try to create an array inside an array, and that is not going to be ok in the E2).
The localized `luaTableToE2` has fixed this by using a number instead of boolean, to represent a depth at which it will enable array optimization.
So, the depth/level of `0` means to enable array optimization on and past the second iteration :)

Cheatoid commented:
*cough* the `try` function is currently broken because it relies on the broken `vex.luaTableToE2` 😭
I'll leave that for you to fix the `vex.luaTableToE2` function, cuz I ain't touching that file, not touching it even with 5 foot stick.
`try` was working fine in the beginning tho (like after the initial patch commit or something like that) and now it is broken, i am not sure at which commit it broke up.

-----

Nonetheless, I have just quick-tested everything from the Self-Aware2 (SA2), which is working great at this point.
I am not going to touch anything anymore in SA2. I would say that's it for this PR.
Pull the latest commit and give it a try. Can you confirm SA2 is fully stable as well?
Can we make a final review now?

Vurv78 commented:
Back, I can put my full attention to this since I've dealt with my school stuff, I shouldn't have broken anything that selfaware2 depends on, but I will leave this pr here until we can both confirm it's fine

Vurv78 commented:
```lua
local test_table = {
    1,"hello",{1,2,3,4,5},"cool",{6,9,6,9,{1,2,3,4,5}}
}
test_table.tool = 400
test_table.self = test_table

-- The second arg being true checks for valid tables that could fit as e2 arrays.
local E2Tbl = vex.luaTableToE2(test_table,true)

PrintTable(E2Tbl)
print(E2Tbl.s.self == E2Tbl) --> true

print( vex.getE2Type( {1,2,3}, true) ) --> array

print( vex.getE2Type( {1,2,3,{}}, true) ) --> table

```
Tests featuring the improved array inferrence

Vurv78 commented:
Need to update printGlobal, it's broken now..

Vurv78 commented:
Tested all of your test code that you posted btw, it all works perfectly

Cheatoid commented on lua/vex_library/modules/server/e2helperfuncs.lua:
```diff
@@ -3,7 +3,7 @@
 local printf = vex.printf
 local isbool, isnumber, isstring, isentity, isangle, isvector, istable, IsColor, isfunction, getmetatable, pcall, IsValid = isbool, isnumber, isstring, isentity, isangle, isvector, istable, IsColor, isfunction, getmetatable, pcall, IsValid
-local string_upper, table_copy, table_IsEmpty, table_IsSequential = string.upper, table.Copy, table.IsEmpty, table.IsSequential
+local string_upper, table_copy, table_IsSequential = string.upper, table.Copy, table.IsEmpty, table.IsSequential
```
Bug: The left hand side defines 3 variables and the right hand side has 4.
Notice, `table_IsSequential` is defined as `table.IsEmpty`.

Cheatoid commented on lua/vex_library/modules/server/e2helperfuncs.lua:
```diff
@@ -16,20 +17,127 @@ vex.registerExtension = function(name,enabled,helpstr,...)
-                return "ARRAY"
+vex.registerConstant = function(name, value)
+    -- This exists because E2Lib doesn't return a passed value, which can be handy (see selfaware2).
+    E2Lib.registerConstant(string.upper(name), value) -- Upper shouldn't be used a lot
```
There's no need to call `string.upper` here, the E2Lib will do that.

Cheatoid replied:
https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/init.lua#L166

Cheatoid commented:
Oh look what I found:
https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/e2lib.lua#L231-L251

Vurv78 commented:
Lol, would be nice if we found this earlier, even if it is pretty much useless in comparison to how in-depth this type-guessing is in comparison

Vurv78 commented:
Hmm.. that guess function is only used in the e2helper actually :/
Yeah would've been useless if we found it then..
[5] is the inverse of [6] btw, I don't think there's really a difference except 6 needs to be pcalled and has more specific conditions

Cheatoid commented:
Indeed yeah, our does better job. 😄
Except, we don't have type guessing for the `ranger` (`xrd`) type, which can easily be added under `istable` check. (see if there is `HitPos` field).
Also, not sure why you got rid of Vector2 optimization?

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_vex_main.lua:
```diff
@@ -7,54 +7,38 @@
+        endpos = Vector(endpos[1],endpos[2],endpos[3]),
+        filter = cleanupTable(filt)
+    }
+    return tr and luaTableToE2(tr) or newE2Table()
```
If you do add type-guessing for `ranger`, then this E2 function should be changed to return `ranger` instead of `table`.

Cheatoid commented:
Here's the quick test code for the `rangerOffsetManual` E2 function:
[See gist here](https://gist.github.com/Cheatoid/2e3dd9802fb0153dac46f09f2dc7a0b2).

Cheatoid commented on lua/entities/gmod_wire_expression2/core/custom/sv_rune2.lua:
```diff
@@ -2,24 +2,6 @@ local table_remove = table.remove
 
-
--- Builds a body to run an e2 udf and pass args to it.
-local function buildBody(args) -- Why does the wireteam do this
```
Shouldn't have removed this. I left this there because for the vararg support for the `try` in the next PR.

Vurv78 replied:
I didn't remove it, it's in vex now

Vurv78 commented:
Now it has support for rangerdata in the type inference..
Now this should be good?? At this point even if there are bugs I just want to pull this..

Vurv78 commented:
I feel like next PR I want to ditch the type inference as a whole, we can just use the funcs_ret table of udfs specifically for the 'try' function and use a variable melding function or something to get the value of the lua variable in e2 form

Cheatoid commented on lua/vex_library/modules/server/e2helperfuncs.lua:
```diff
@@ -140,6 +137,25 @@ vex.luaTableToE2 = function(tbl,checkForArrays)
+-- Moved from rune2
+-- Builds a body to run an e2 udf and pass args to it.
+-- TODO: Probably want to use type inferrence with this so we don't have to provide types in the table (and could turn this into a vararg function)
+vex.buildBody = function(args)
```
Ah it is in the helpers now. Ok.

Cheatoid commented:
I haven't tested printGlobal myself tho. So not sure about it.
Other than that, it looks good to go. `try` will be fixed within the next PR.

Vurv78 commented:
Yeah, printGlobal works fine for me right now, I'll pull then, thanks for contributing!
